inc/az_ulib_registry_api.h

```diff
@@ -0,0 +1,198 @@
+ *
+ * @param[in]   key                 The #az_span key to add to the registry.
+ * @param[in]   value               The #az_span value to add to the registry.
+ *
+ * @pre         Registry shall already be initialized.
+ * @pre         \p key              shall not be `#AZ_SPAN_EMPTY`.
+ * @pre         \p value            shall not be `#AZ_SPAN_EMPTY`.
+ *
+ * @return The #az_result with the result of the registry operations.
+ *      @retval #AZ_OK                            If adding a value to the registry was successful.
+ *      @retval #AZ_ERROR_ULIB_ELEMENT_DUPLICATE  If there is a key within the registry that is the
+ *                                                same as the new key.
+ *      @retval #AZ_ERROR_ULIB_SYSTEM             If the `az_ulib_registry_add` operation failed on
+ *                                                the system level.
+ *      @retval #AZ_ERROR_ULIB_BUSY               If the resources necessary for the
+ *                                                `az_ulib_registry_add` operation are busy.
+ *      @retval #AZ_ERROR_OUT_OF_MEMORY           If the flash space for `az_ulib_registry_add`
+ *                                                is not enough for a new registry entry.
+ */
+AZ_NODISCARD az_result az_ulib_registry_add(az_span key, az_span value);
+
+/**
+ * @brief   This function removes an #az_span key and its corresponding #az_span value from the
+ * device registry.
+ *
+ * This function goes through the registry and marks the node with the given key value pair as
+ * deleted so that it can no longer be accessible.
+ *
+ * @param[in]   key                 The #az_span key to remove from the registry.
+ *
+ * @pre         Registry shall already be initialized.
+ * @pre         \p key              shall not be `#AZ_SPAN_EMPTY`.
+ *
+ * @return The #az_result with the result of the registry operations.
+ *      @retval #AZ_OK                            If removing a value from the registry was
+ *                                                successful.
+ *      @retval #AZ_ERROR_ULIB_SYSTEM             If the `az_ulib_registry_delete` operation failed
+ *                                                on the system level.
+ *      @retval #AZ_ERROR_ULIB_BUSY               If the resources necessary for the
+ *                                                `az_ulib_registry_delete` operation are busy.
+ */
+AZ_NODISCARD az_result az_ulib_registry_delete(az_span key);
+
+/**
+ * @brief   This function initializes the device registry.
+ *
+ * This function initializes components that the registry needs upon reboot. This function is not
+ * thread safe and all other APIs shall only be invoked after the initialization ends.
+ *
+ * @param[in]   registry_cb         The pointer to #az_ulib_registry_control_block with the control
+ *                                  block that contains the registry memory.
+ *
+ * @pre         Registry shall **not** be initialized.
```

Comment:
shouldn't this return an error if the registry has already been initialized? Also do we allow multiple registries to be initialized? For instance:
```c
const az_ulib_registry_control_block registry_cb1;
const az_ulib_registry_control_block registry_cb2;

az_ulib_registry_init(&registry_cb1);
az_ulib_registry_init(&registry_cb2);
```

Reply:
This is a good point.

Registry is a singleton class, double initialization is a developers error, so we stop the code using the precondition check, not returning an error. However, in release mode, this function will change it's behavior, overwriting the singleton instance. 

Because this API is used only in the DCF-baseimage, I don't think we should add any extra code for that check, but we can discuss more about it.

Reply:
After chat offline, we decided to keep the Precondition only and add some more documentation about it.

Reply:
#Pedning